app/services/recaptcha_validator.rb

```diff
@@ -0,0 +1,45 @@
+class RecaptchaValidator
+  class ValidationError < StandardError; end
+
+  attr_reader :score_threshold
+
+  def initialize(score_threshold: 0.0)
+    @score_threshold = score_threshold
+  end
+
+  def exempt?
+    !score_threshold.positive?
+  end
+
+  def valid?(recaptcha_token)
+    return true if exempt?
+
+    response = faraday.post(
+      VERIFICATION_ENDPOINT,
+      URI.encode_www_form(
+        secret: IdentityConfig.store.recaptcha_secret_key,
+        response: recaptcha_token,
+      ),
+    )
+
+    if !response.body['success']
+      raise ValidationError.new("reCAPTCHA validation error: #{response.body['error-codes']}")
+    end
+
+    response.body['score'] >= score_threshold
+  rescue ValidationError => error
+    NewRelic::Agent.notice_error(error)
+    true
+  end
+
+  private
+
+  def faraday
+    Faraday.new do |conn|
+      conn.request :instrumentation, name: 'request_log.faraday'
+      conn.response :json
+    end
+  end
+
+  VERIFICATION_ENDPOINT = 'https://www.google.com/recaptcha/api/siteverify'.freeze
```

Comment:
as a style/convention, I'd move this to the top of the file

Reply:
Oops, yeah, I think I meant for it to be under `private`, but still otherwise above any defined methods. Would that track with what you'd expect styling-wise?

Reply:
constants are public in ruby even if they're below `private` 😭 , you'd need to flag with [`private_constant`](https://ruby-doc.org/core-2.5.3/Module.html#method-i-private_constant), and IMO visibility in ruby is a joke so I wouldn't worth bothering

sometimes if a constant is just for reference, I'll define it right above the method that uses it? but otherwise generally put them at the top of the file

Reply:
Moved in c9d403411.